source-mysql/throughput_test.go

```diff
@@ -0,0 +1,202 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"math/rand"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/estuary/connectors/sqlcapture"
+	"github.com/estuary/connectors/sqlcapture/tests"
+	"github.com/estuary/flow/go/protocols/airbyte"
+	"github.com/google/uuid"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+// The benchmarks 'BenchmarkBackfill' and 'BenchmarkReplication' initialize
+// several tables with a bunch of synthetic data and benchmark how long it
+// takes to capture the data (via Backfill and Replication, respectively).
+//
+// In both cases the tests will run one capture, spreading the data across
+// 10 tables, with the benchmark 'N' parameter controlling the number of
+// rows in each table. To be precise, the number of rows is set to N*100,
+// so each 'op' in the 'ns/op' measurement reflects the marginal cost of
+// 1,000 row-captures.
+//
+// These benchmarks have a rather high setup cost, so the default `benchtime`
+// target of one second isn't nearly enough. Aim for at least 30 seconds to
+// get useful numbers (and 300 seconds for perfect accuracy):
+//
+//     $ LOG_LEVEL=warn go test -run=None -bench=. -benchtime=30s ./source-mysql/
+//     $ LOG_LEVEL=warn go test -run=None -bench=. -benchtime=30s -memprofile memprofile.out -cpuprofile profile.out ./source-mysql/
+//
+
+func BenchmarkBackfill(b *testing.B) { benchmarkBackfills(b, 1, 10, b.N*100) }
+
+func BenchmarkReplication(b *testing.B) { benchmarkReplication(b, 1, 10, b.N*100) }
+
+func benchmarkBackfills(b *testing.B, iterations, numTables, rowsPerTable int) {
+	b.StopTimer()
+	b.ResetTimer()
+
+	// Set up multiple tables full of data
+	logrus.WithFields(logrus.Fields{
+		"rows":         rowsPerTable * numTables,
+		"rowsPerTable": rowsPerTable,
+		"tables":       numTables,
+	}).Info("initializing tables")
+
+	var tb, ctx = TestBackend, context.Background()
+	var tables []string
+	var grp errgroup.Group
+	for i := 0; i < numTables; i++ {
+		var table = tb.CreateTable(ctx, b, fmt.Sprintf("table%d", i), "(id INTEGER PRIMARY KEY, uid TEXT, name TEXT, status INTEGER, modified DATE, foo_id INTEGER, padding TEXT)")
+		tables = append(tables, table)
+		//grp.Go(func() error { return populateTable(ctx, b, tb, table, rowsPerTable) })
+		populateTable(ctx, b, tb, table, rowsPerTable)
+	}
+	if err := grp.Wait(); err != nil {
```

Comment:
Depending on if the `grp.Go` above is supposed to be used or not, this may or may not be necessary.